### ISSUE TITLE
SDK should define Python dependencies with version ranges

### ISSUE THREAD

**Issue by AniX** [pulumi/pulumi#14307]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

I'd like to suggest to improve the user experience for Python users of the SDK to prevent issues like https://github.com/pulumi/pulumi/issues/14258 and https://github.com/pulumi/pulumi/issues/13738.

I wonder if it would be worth revisiting the Python dependency configuration, as version conflicts will keep popping up, especially if users have also other dependencies that may need the same dependencies (pulumi defines some very popular libraries, so conflicts are likely).

Talking from my own experience, Pulumi SDK should define only the minimum required versions it actually needs to work. That means that it should have been 1.56.0, not 1.56.2, as I doubt that it makes any difference for Pulumi whether 1.56.2, 1.56.1, or 1.56.0 is installed. And with semantic versioning you can typically rely on the fact that newer versions 1.57.x, 1.58.x will not break pulumi, so they shouldn't be excluded by default.

Assuming the SDK actually requires features of 1.56.0, the dependency should be defined as:

```
grpcio>=1.56.0,<2.0.0
```

Equivalent in poetry (poetry would also avoid out-of-sync dependencies between `setup.py` and `requirements.txt`):

```
grpcio = "^1.56.0"
```

I admit it is a little optimistic that grpcio and all other dependencies will never violate semantic versioning. It may still happen that a future grpcio 1.x.x breaks existing functionality - but even if it would happen:
1. the team can either patch pulumi code to remediate the breaking change
2. or the team can exclude the breaking version in a pulumi patch release.
3. Finally, users don't need to rely on a pulumi patch, because they can still explicitly pin to lower versions on their end if needed and while these versions are still greater than the minimum version pulumi requires.

I believe this would give Python users a much greater flexibility and reliability in how they use pulumi. And since pulumi SDK doesn't have that many dependencies, I think it wouldn't take much time away from feature work etc.

### Affected area/feature

Pulumi SDK's `setup.py` and `requirements.txt` - alternatively to `pyproject.toml` for poetry.

**Comment by Frassle** [comment]
We've historically had a few breakages from grpcio releasing bad versions. At some point we made the decision that most of our users would be better off with grpcio locked to a version that had gone through our CI system and been tested.
With https://github.com/pulumi/pulumi/issues/14258 we've decided to go back to trying to use a minimum dependency version instead, see https://github.com/pulumi/pulumi/pull/14259. We'll see how that goes.
